src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -908,8 +936,162 @@ - (void)readAttributesWithEndpointID:(NSNumber * _Nullable)endpointID
+}
+
+- (void)readWithAttributePaths:(NSArray<MTRAttributeRequestPath *> * _Nullable)attributePaths
+                    EventPaths:(NSArray<MTREventRequestPath *> * _Nullable)eventPaths
+                        params:(MTRReadParams * _Nullable)params
+                         queue:(dispatch_queue_t)queue
+                    completion:(MTRDeviceResponseHandler)completion
+{
+    if ((attributePaths == nil || [attributePaths count] == 0) && (eventPaths == nil || [eventPaths count] == 0)) {
+        dispatch_async(queue, ^{
+            completion(nil, [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT]);
+        });
+        return;
+    }
+
+    attributePaths = (attributePaths == nil) ? nil : [attributePaths copy];
+    eventPaths = (eventPaths == nil) ? nil : [eventPaths copy];
+    params = (params == nil) ? nil : [params copy];
+    auto * bridge = new MTRDataValueDictionaryCallbackBridge(queue, completion,
+        ^(ExchangeManager & exchangeManager, const SessionHandle & session, MTRDataValueDictionaryCallback successCb,
+            MTRErrorCallback failureCb, MTRCallbackBridgeBase * bridge) {
+            // interactionStatus tracks whether the whole read interaction has failed.
+            //
+            // Make sure interactionStatus survives even if this block scope is destroyed.
+            auto interactionStatus = std::make_shared<CHIP_ERROR>(CHIP_NO_ERROR);
+
+            auto resultArray = [[NSMutableArray alloc] init];
+            auto onAttributeSuccessCb = [resultArray](const app::ConcreteClusterPath & clusterPath, const AttributeId aAttributeId,
+                                            const MTRDataValueDictionaryDecodableType & aData) {
+                app::ConcreteAttributePath attribPath(clusterPath.mEndpointId, clusterPath.mClusterId, aAttributeId);
+                [resultArray addObject:@ {
+                    MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:attribPath],
+                    MTRDataKey : aData.GetDecodedObject()
+                }];
+            };
+
+            auto onEventSuccessCb = [resultArray](const app::ConcreteClusterPath & clusterPath, const EventId aEventId,
+                                        const MTRDataValueDictionaryDecodableType & aData) {
+                app::ConcreteEventPath eventPath(clusterPath.mEndpointId, clusterPath.mClusterId, aEventId);
+                [resultArray addObject:@ {
+                    MTREventPathKey : [[MTREventPath alloc] initWithPath:eventPath],
+                    MTRDataKey : aData.GetDecodedObject()
+                }];
+            };
+
+            auto onFailureCb = [resultArray, interactionStatus](
+                                   const app::ConcreteClusterPath * clusterPath, const uint32_t aValueId, CHIP_ERROR aError) {
+                if (clusterPath) {
+                    app::ConcreteAttributePath attribPath(clusterPath->mEndpointId, clusterPath->mClusterId, aValueId);
```

Comment:
How do you know this is the attribute case?  This could be an event.  Seems like we need two different error failure callbacks, like we have two different success callbacks.... Or possibly three: one for path-specific attribute errors, one for path-specific event errors, one for non-path-specific errors.  That way in the non-path-specific case the caller doesn't have to guess whether to call the "event" or "attribute" callback.

Reply:
Oh.... I will be modified.